src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+    //  size, the offset of the last object (chunk size * highest ID seen)
+    //  and the actual size (offset of last object + size of highest ID seen)
+    //
+    //  This logic doesn't take account of striping.
+    uint64_t inode_no = 0;
+    uint64_t obj_id = 0;
+    r = parse_oid(oid, &inode_no, &obj_id);
+    if (r != 0) {
+      dout(4) << "Bad object name '" << oid << "' skipping" << dendl;
+      continue;
+    }
+
+    // Generate 0th object name, where we will accumulate sizes/mtimes
+    object_t zeroth_object = InodeStore::get_object_name(inode_no, frag_t(), "");
+
+    // Construct a librados operation invoking our class method
+    librados::ObjectReadOperation op;
+    bufferlist inbl;
+    ::encode(std::string("scan_size"), inbl);
+    ::encode(std::string("scan_mtime"), inbl);
+    ::encode(obj_id, inbl);
+    ::encode(size, inbl);
+    ::encode(mtime, inbl);
+
+    op.exec("cephfs", "accumulate_inode_metadata", inbl);
+
+    bufferlist outbl;
+    int r = data_io.operate(zeroth_object.name, &op, &outbl);
+    if (r < 0) {
+      derr << "Failed to store size data from '"
+        << oid << "' to '" << zeroth_object.name << "': "
+        << cpp_strerror(r) << dendl;
+      continue;
+    }
+  }
+
+  return 0;
+}
+
+/*
+ * Other handy commands:
+ *  A Find inode by backtrace expression, e.g. "myimportantfile"
+ *  B Selectively recover only particular inodes, e.g. those identified
+ *    by A.
+ *  C When recovering inodes, write files out to local disk instead of
+ *    trying to inject them into the CephFS metadata pool.
+ */
+
+int DataScan::recover()
+{
+#ifdef SHARDED_PGLS
+  float progress = 0.0;
+  librados::NObjectIterator i = data_io.nobjects_begin(n, m);
+#else
+  librados::NObjectIterator i = data_io.nobjects_begin();
+#endif
+  librados::NObjectIterator i_end = data_io.nobjects_end();
+
+  bool roots_present;
+  int r = driver->check_roots(&roots_present);
+  if (r != 0) {
+    derr << "Unexpected error checking roots: '"
+      << cpp_strerror(r) << "'" << dendl;
+    return r;
+  }
+
+  if (!roots_present) {
+    std::cerr << "Some or all system inodes are absent.  Run 'init' from "
+      "one node before running 'recover'" << std::endl;
+    return -EIO;
+  }
+
+  for (; i != i_end; ++i) {
+    const std::string oid = i->get_oid();
+#ifdef SHARDED_PGLS
+    if (i.get_progress() != progress) {
+      if (int(i.get_progress() * 100) / 5 != int(progress * 100) / 5) {
+        std::cerr << percentify(i.get_progress()) << "%" << std::endl;
+      }
+      progress = i.get_progress();
+    }
+#endif
+
+    uint64_t obj_name_ino = 0;
+    uint64_t obj_name_offset = 0;
+    r = parse_oid(oid, &obj_name_ino, &obj_name_offset);
+    if (r != 0) {
+      dout(4) << "Bad object name '" << oid << "', skipping" << dendl;
+      continue;
+    }
+
+    // We are only interested in 0th objects during this phase: we touched
+    // the other objects during recover_extents
```

Comment:
Rather than filter on the client, we should add to the pgls_filter mechanism (or extend it to be more awesome!) that already exists. That way the OSDs can filter for us and send less irrelevant data back over the wire.